### PR TITLE
Unilogin api

### PR DESCRIPTION
**Pull request by rimi-itk**
https://jira.itkdev.dk/browse/ER-682

### CODE REVIEW

File: ding_unilogin/ding_unilogin.test [hunk 1]
```diff
@@ -0,0 +1,160 @@
+<?php
+
```
Comment (cableman) [inline ding_unilogin/ding_unilogin.test]: Technical all the new files is missing `@file` documentation.

File: ding_unilogin/src/Controller/InstitutionController.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+use Drupal\ding_unilogin\Exception\HttpNotFoundException;
+use Drupal\ding_unilogin\Exception\HttpUnauthorizedException;
+
+/**
+ * Institution controller.
+ */
+class InstitutionController {
+
+  /**
+   * Handle request.
```
[the remark this answers is not in the document]
Comment (cableman) [reply ding_unilogin/src/Controller/InstitutionController.php]: Missing `@throws` documentation for invalid users auth.

File: ding_unilogin/src/Controller/InstitutionController.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+      if ('POST' === $method) {
+        if (variable_get('ding_unilogin_api_token_write') !== $token) {
+          throw new HttpUnauthorizedException();
+        }
+
+        return $this->update();
+      }
+
+      if (variable_get('ding_unilogin_api_token_read') !== $token) {
```
Comment (cableman) [inline ding_unilogin/src/Controller/InstitutionController.php]: So PUT, PATCH or DELETE request would get the read access (if valid token)? A `not supported` http response might be more correct.

File: ding_unilogin/src/Controller/InstitutionController.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      return ['data' => $institutions[$id]];
+    }
+
+    throw new HttpNotFoundException(sprintf('Invalid institution id: %s', $id));
+  }
+
+  /**
+   * Update the list of institutions.
+   */
+  public function update() {
```
Comment (cableman) [inline ding_unilogin/src/Controller/InstitutionController.php]: Is this not more a `create` as it comes through a `POST` request and overrides all the institutions? Should an `PATCH` have been create to update a single institutions in the list?

File: ding_unilogin/src/Controller/InstitutionController.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    if (empty($payload) || empty($payload['data'])) {
+      throw new HttpBadRequestException(sprintf('Invalid content'));
+    }
+
+    $data = $payload['data'];
+
+    // Validate data.
+    foreach ($data as $id => &$item) {
+      if (!isset($item['name'], $item['group'], $item['type'], $item['number_of_members'])
+        || !preg_match('/^[a-z0-9]{6}$/i', (string) $id)) {
```
Comment (cableman) [inline ding_unilogin/src/Controller/InstitutionController.php]: Only validation of the content of the ID and not the group, type or number_of_members (is an number).

File: ding_unilogin/ding_unilogin.info [hunk 6]
```diff
@@ -3,9 +3,12 @@ description = Allow login via UNI•Login.
 core = 7.x
 package = Ding!
 dependencies[] = ding_user
+dependencies[] = xautoload
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply ding_unilogin/ding_unilogin.info]: As I understand it, `xautoload` should be a dependency to get stuff in the module's `src` folder loaded.

This module does not depend on `ding_test`, but in order to load tests from other `ding` modules, the `ding_test` module must be installed.

File: ding_unilogin/ding_unilogin.module [hunk 7]
```diff
@@ -51,6 +63,13 @@ function ding_unilogin_menu() {
     'type' => MENU_CALLBACK,
   );
 
+  $items['unilogin/api/%'] = array(
+    'page callback' => 'ding_unilogin_api_router',
+    'page arguments' => array(2),
+    'access callback' => TRUE,
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply ding_unilogin/ding_unilogin.module]: Access checks are performed in the api controller.

File: ding_unilogin/src/Controller/InstitutionController.php (hunk 3 above)
Comment (rimi-itk) [reply ding_unilogin/src/Controller/InstitutionController.php]: Good point!

File: ding_unilogin/src/Controller/InstitutionController.php (hunk 4 above)
Comment (rimi-itk) [reply ding_unilogin/src/Controller/InstitutionController.php]: It is an update (actually an overwrite via `POST`). We don't need to be able to update a single institution (for now).

File: ding_unilogin/src/Controller/InstitutionController.php (hunk 5 above)
Comment (rimi-itk) [reply ding_unilogin/src/Controller/InstitutionController.php]: Nice catch! We really only care about `id` and `number_of_members` – I've added better checks.

File: ding_unilogin/ding_unilogin.module [hunk 8]
```diff
@@ -43,6 +46,15 @@ function ding_unilogin_menu() {
     'file' => 'ding_unilogin.admin.inc',
   );
 
+  $items['admin/config/ding/unilogin_api'] = array(
+    'title' => 'UNI•Login api',
+    'description' => 'Configure UNI•Login api',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_unilogin_admin_api_form'),
+    'access arguments' => array('configure unilogin'),
+    'file' => 'ding_unilogin.admin.inc',
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply ding_unilogin/ding_unilogin.module]: Moved!

File: ding_unilogin/ding_unilogin.test (hunk 1 above)
Comment (rimi-itk) [reply ding_unilogin/ding_unilogin.test]: No: `files that contain a namespaced class/interface/trait, whose file name is the class name with a .php extension, and whose file path is closely related to the namespace (under PSR-4 or a similar standard), SHOULD NOT have a @file documentation block.` (cf. https://www.drupal.org/docs/develop/standards/api-documentation-and-comment-standards#file)

File: ding_unilogin/ding_unilogin.module [hunk 9]
```diff
@@ -596,3 +615,66 @@ function _ding_unilogin_set_user_data(array $data = NULL) {
     $_SESSION['ding_unilogin_user_data'] = $data;
   }
 }
+
+/**
+ * Get UNI•Login institutions.
```
[the remark this answers is not in the document]
Comment (cableman) [reply ding_unilogin/ding_unilogin.module]: Still missing return type.

File: ding_unilogin/ding_unilogin.module [hunk 10]
```diff
@@ -596,3 +615,66 @@ function _ding_unilogin_set_user_data(array $data = NULL) {
...
+ */
+function _ding_unilogin_get_institutions() {
+  return variable_get('ding_unilogin_institutions', []);
+}
+
+/**
+ * Set UNI•Login institutions.
```
[the remark this answers is not in the document]
Comment (cableman) [reply ding_unilogin/ding_unilogin.module]: Still missing return type.

File: ding_unilogin/ding_unilogin.module [hunk 11]
```diff
@@ -596,3 +615,66 @@ function _ding_unilogin_set_user_data(array $data = NULL) {
...
+ */
+function _ding_unilogin_set_institutions(array $institutions) {
+  variable_set('ding_unilogin_institutions', $institutions);
+}
+
+/**
+ * Api router.
```
[the remark this answers is not in the document]
Comment (cableman) [reply ding_unilogin/ding_unilogin.module]: Still missing return type.

File: ding_unilogin/src/Exception/HttpException.php [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Class HttpException.
+ *
+ * @package Drupal\ding_unilogin\Exception
+ */
+abstract class HttpException extends \RuntimeException {
+
+  /**
+   * Get http status header including status code.
```
[the remark this answers is not in the document]
Comment (cableman) [reply ding_unilogin/src/Exception/HttpException.php]: Still missing return type.

**Pull request merged**